src/messaging/ExchangeMgr.cpp

```diff
@@ -0,0 +1,349 @@
+#endif
+
+#ifndef __STDC_LIMIT_MACROS
+#define __STDC_LIMIT_MACROS
+#endif
+
+#include <inttypes.h>
+#include <stddef.h>
+
+#include <core/CHIPCore.h>
+#include <core/CHIPEncoding.h>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+#include <protocols/CHIPProtocols.h>
+#include <protocols/common/CommonProtocol.h>
+#include <protocols/security/CHIPSecurity.h>
+#include <support/CHIPFaultInjection.h>
+#include <support/CodeUtils.h>
+#include <support/RandUtils.h>
+#include <support/logging/CHIPLogging.h>
+#include <system/SystemStats.h>
+#include <system/SystemTimer.h>
+
+using namespace chip::Encoding;
+using namespace chip::Inet;
+using namespace chip::System;
+
+namespace chip {
+
+/**
+ *  Constructor for the ExchangeManager class.
+ *  It sets the state to kState_NotInitialized.
+ *
+ *  @note
+ *    The class must be initialized via ExchangeManager::Init()
+ *    prior to use.
+ *
+ */
+ExchangeManager::ExchangeManager()
+{
+    mState = State::kState_NotInitialized;
+}
+
+CHIP_ERROR ExchangeManager::Init(SecureSessionMgrBase * sessionMgr)
+{
+    if (mState != State::kState_NotInitialized)
+        return CHIP_ERROR_INCORRECT_STATE;
+
+    mSessionMgr = sessionMgr;
+
+    mNextExchangeId = GetRandU16();
+
+    memset(ContextPool, 0, sizeof(ContextPool));
+    mContextsInUse = 0;
+
+    memset(UMHandlerPool, 0, sizeof(UMHandlerPool));
+    OnExchangeContextChanged = nullptr;
+
+    sessionMgr->SetDelegate(this);
+
+    mState = State::kState_Initialized;
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ExchangeManager::Shutdown()
+{
+    if (mSessionMgr != nullptr)
+    {
+        mSessionMgr = nullptr;
+    }
+
+    OnExchangeContextChanged = nullptr;
+
+    mState = State::kState_NotInitialized;
+
+    return CHIP_NO_ERROR;
+}
+
+ExchangeContext * ExchangeManager::NewContext(const uint64_t & peerNodeId, void * appState)
+{
+    ExchangeContext * ec = AllocContext();
+
+    if (ec != nullptr)
+    {
+        ec->SetExchangeId(mNextExchangeId++);
+        ec->SetPeerNodeId(peerNodeId);
+        ec->SetAppState(appState);
+        ec->SetInitiator(true);
+        ChipLogProgress(ExchangeManager, "ec id: %d, AppState: 0x%x", (ec - ContextPool + 1), ec->GetAppState());
+    }
+
+    return ec;
+}
+
+ExchangeContext * ExchangeManager::FindContext(uint64_t peerNodeId, void * appState, bool isInitiator)
+{
+    ExchangeContext * ec = ContextPool;
+
+    for (int i = 0; i < CHIP_CONFIG_MAX_EXCHANGE_CONTEXTS; i++, ec++)
+    {
+        if (ec->GetExchangeMgr() != nullptr && ec->GetPeerNodeId() == peerNodeId && ec->GetAppState() == appState &&
+            ec->IsInitiator() == isInitiator)
+            return ec;
+    }
+
+    return nullptr;
+}
+
+CHIP_ERROR ExchangeManager::RegisterUnsolicitedMessageHandler(uint32_t protocolId, ExchangeContext::MessageReceiveFunct handler,
+                                                              void * appState)
+{
+    return RegisterUMH(protocolId, (int16_t) -1, handler, appState);
```

Comment:
A constant for 'kAnyMessageType' would be useful (and would not require casts)

Reply:
Acked